File: implementation/swift/GameEventScriptConformance/Sources/GameEventScriptConformance/ConformanceRuntimeValueCodec.swift
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

import GameEventScriptRuntime

enum ConformanceExecutionError: Error {
    case invalidInput(String)
}

extension ConformanceData {
    func required(_ key: String) throws -> ConformanceData {
        guard let value = self[key] else { throw ConformanceExecutionError.invalidInput("Missing \(key)") }
        return value
    }

    func text(_ key: String) throws -> String {
        guard let value = self[key]?.stringValue ?? self[key]?.numberValue else {
            throw ConformanceExecutionError.invalidInput("Expected scalar \(key)")
        }
        return value
    }

    func values(_ key: String) -> [ConformanceData] { self[key]?.arrayValue ?? [] }
}

enum ConformanceRuntimeValueCodec {
    static func unit(_ node: ConformanceData) throws -> GesUnit {
        switch node["unit"]?.stringValue?.replacingColonPrefix() {
        case nil, "none": return .none
        case "m", "meter": return .meter
        case "s", "second": return .second
        case "°", "degree": return .degree
        default: throw ConformanceExecutionError.invalidInput("Unknown unit")
        }
    }

    static func number(_ text: String) throws -> Double {
        switch text {
        case "NaN": return .nan
        case "Infinity": return .infinity
        case "-Infinity": return -.infinity
        default:
            guard let value = Double(text) else { throw ConformanceExecutionError.invalidInput("Invalid Binary64") }
            return value
        }
    }

    static func integer(_ text: String) throws -> Int64 {
        guard let value = Int64(text) else { throw ConformanceExecutionError.invalidInput("Invalid Int64") }
        return value
    }

    static func signature(_ node: ConformanceData) throws -> GameEventScriptMessageSignature {
        try GameEventScriptMessageSignature(
            name: node.text("name"), parameters: node.values("parameters").map { $0.stringValue })
    }

    static func message(_ node: ConformanceData) throws -> GameEventScriptMessage {
        let arguments = try node.values("args").map { argument in
            try GameEventScriptMessageArgument(
                name: argument["name"]?.stringValue, value: decode(argument.required("value")))
        }
        return try GameEventScriptMessage(
            name: node.text("name"), arguments: arguments, tags: node.values("tags").compactMap(\.stringValue))
    }

    static func decode(_ node: ConformanceData, mutateSource: Bool = false) throws -> GesValue {
        let type = try node.text("type")
        let unit = try unit(node)
        switch type {
        case ":Nothing": return .nothing
        case ":Boolean": return .boolean(node["value"]?.boolValue == true || node["value"]?.stringValue == "true")
        case ":Number.int64", ":Quantity.int64": return try .integer(integer(node.text("value")), unit: unit)
        case ":Number.binary64", ":Quantity.binary64": return try .float(number(node.text("value")), unit: unit)
        case ":Percentage": return try .percentage(number(node.text("value")))
        case ":Text": return try .text(node.text("value"))
        case ":Tag": return try .tag(node.text("value"))
        case ":Vector":
            return try .vector(
                x: number(node.text("x")), y: number(node.text("y")), z: number(node.text("z")), unit: unit)
        case ":Point":
            return try .point(
                x: number(node.text("x")), y: number(node.text("y")), z: number(node.text("z")), unit: unit)
        case ":List":
            var items = try node.values("items").map { try decode($0) }
            let result = GesValue.list(items)
            if mutateSource && !items.isEmpty { items[0] = .nothing }
            return result
        case ":Dice":
            var rolls = try node.values("rolls").map { value in
                guard let result = Int32(value.numberValue ?? value.stringValue ?? "") else {
                    throw ConformanceExecutionError.invalidInput("Invalid roll")
                }
                return result
            }
            let result = GesValue.dice(rolls)
            if mutateSource && !rolls.isEmpty { rolls[0] = .min }
            return result
        case ":Range.int64":
            return try .integerRange(
                from: integer(node.text("from")), to: integer(node.text("to")), step: integer(node.text("step")))
        case ":Range.binary64":
            return try .floatRange(
                from: number(node.text("from")), to: number(node.text("to")), step: number(node.text("step")))
        case ":Message": return try .message(message(node.required("message")))
        default:
            var entries = try node.values("entries").map {
                try GesMapEntry(key: $0.text("key"), value: decode($0.required("value")))
            }
            let result =
                type == ":Map"
                ? GesValue.map(entries) : GesValue.record(typeName: String(type.dropFirst()), entries: entries)
            if mutateSource && !entries.isEmpty { entries[0] = GesMapEntry(key: "mutated", value: .nothing) }
            return result
        }
    }

    static func binaryEqual(_ left: Double, _ right: Double, comparison: ConformanceData?) -> Bool {
        if left.isNaN || right.isNaN { return left.isNaN && right.isNaN }
        if comparison?["binary64"]?["mode"]?.stringValue != "ulp" { return left.bitPattern == right.bitPattern }
        if left == right { return true }
        if !left.isFinite || !right.isFinite { return false }
        let mask: UInt64 = 0x8000_0000_0000_0000
        func ordered(_ number: Double) -> UInt64 {
            number.bitPattern & mask == 0 ? number.bitPattern | mask : ~number.bitPattern
        }
        let a = ordered(left)
        let b = ordered(right)
        let limit = UInt64(comparison?["binary64"]?["maxUlps"]?.numberValue ?? "0") ?? 0
        return (a >= b ? a - b : b - a) <= limit
    }

    static func messagesEqual(
        _ expected: ConformanceData, _ actual: GameEventScriptMessage, comparison: ConformanceData?
    ) throws -> Bool {
        guard scalarEqual(try expected.text("name"), actual.name) else { return false }
        let tags = expected.values("tags").compactMap(\.stringValue)
        guard tags.count == actual.tags.count, zip(tags, actual.tags).allSatisfy({ scalarEqual($0, $1) }) else {
            return false
        }
        let args = expected.values("args")
        guard args.count == actual.arguments.count else { return false }
        for (index, argument) in args.enumerated() {
            guard scalarEqual(argument["name"]?.stringValue ?? "_", actual.arguments.nameAt(index)) else {
                return false
            }
            if try !equal(argument.required("value"), actual.arguments[index], comparison: comparison) { return false }
        }
        return true
    }

    // Compare authored transport kinds before invoking any normalizing Runtime factory.
    static func equal(_ expected: ConformanceData, _ actual: GesValue, comparison: ConformanceData?) throws -> Bool {
        let type = try expected.text("type")
        let expectedUnit = try unit(expected)
        switch type {
        case ":Nothing": return actual.isNothing
        case ":Boolean": return actual.kind == .boolean && (expected["value"]?.boolValue == true) == actual.asBoolean
        case ":Number.int64", ":Quantity.int64":
            return try actual.kind == .integer && expectedUnit == actual.unit
                && integer(expected.text("value")) == actual.integerValue
        case ":Number.binary64", ":Quantity.binary64", ":Percentage":
            let number = try number(expected.text("value"))
            if number.isNaN { return actual.isNothing }
            return actual.kind == (type == ":Percentage" ? .percentage : .float) && expectedUnit == actual.unit
                && binaryEqual(number, actual.asNumber, comparison: comparison)
        case ":Text", ":Tag":
            return try actual.kind == (type == ":Text" ? .text : .tag)
                && scalarEqual(expected.text("value"), actual.textValue ?? "")
        case ":Vector", ":Point":
            let x = try number(expected.text("x"))
            let y = try number(expected.text("y"))
            let z = try number(expected.text("z"))
            if x.isNaN || y.isNaN || z.isNaN { return actual.isNothing }
            guard actual.kind == (type == ":Vector" ? .vector : .point), expectedUnit == actual.unit,
                let spatial = actual.spatialValue
            else { return false }
            return binaryEqual(x, spatial.x, comparison: comparison)
                && binaryEqual(y, spatial.y, comparison: comparison)
                && binaryEqual(z, spatial.z, comparison: comparison)
        case ":List":
            guard let items = actual.listValue, items.count == expected.values("items").count else { return false }
            for (a, b) in zip(expected.values("items"), items) {
                if try !equal(a, b, comparison: comparison) { return false }
            }
            return true
        case ":Dice":
            let rolls = try expected.values("rolls").map { value in
                guard let roll = Int32(value.numberValue ?? value.stringValue ?? "") else {
                    throw ConformanceExecutionError.invalidInput("Invalid expected roll")
                }
                return roll
            }
            return actual.diceRolls == rolls
        case ":Range.int64":
            guard let range = actual.integerRangeValue else { return false }
            return try range.from == integer(expected.text("from")) && range.to == integer(expected.text("to"))
                && range.step == integer(expected.text("step"))
        case ":Range.binary64":
            let from = try number(expected.text("from"))
            let to = try number(expected.text("to"))
            let step = try number(expected.text("step"))
            if from.isNaN || to.isNaN || step.isNaN { return actual.isNothing }
            guard let range = actual.floatRangeValue else { return false }
            return binaryEqual(from, range.from, comparison: comparison)
                && binaryEqual(to, range.to, comparison: comparison)
                && binaryEqual(step, range.step, comparison: comparison)
        case ":Message":
            guard let message = actual.messageValue else { return false }
            return try messagesEqual(expected.required("message"), message, comparison: comparison)
        default:
            guard actual.kind == (type == ":Map" ? .map : .record), let entries = actual.mapEntries else {
                return false
            }
            if type != ":Map" && !scalarEqual(String(type.dropFirst()), actual.customTypeName ?? "") { return false }
            var normalized: [([UInt32], ConformanceData)] = []
            for entry in expected.values("entries") {
                let key = try entry.text("key").unicodeScalars.map(\.value)
                normalized.removeAll { $0.0 == key }
                normalized.append((key, try entry.required("value")))
            }
            guard normalized.count == entries.count else { return false }
            for (key, value) in normalized {
                guard let match = entries.first(where: { $0.key.unicodeScalars.map(\.value) == key }),
                    try equal(value, match.value, comparison: comparison)
                else { return false }
            }
            return true
        }
    }

    static func scalarEqual(_ a: String, _ b: String) -> Bool { a.unicodeScalars.elementsEqual(b.unicodeScalars) }
}

extension String {
    fileprivate func replacingColonPrefix() -> String { hasPrefix(":") ? String(dropFirst()) : self }
}

File: implementation/csharp/GameEventScript.Conformance/src/ConformanceRuntimeValueCodec.cs
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections.Generic;
using System.Globalization;
using GameEventScript.Api;
using GameEventScript.Runtime.Values;
using static GameEventScript.Api.GameEventScriptBytecodeTypeKind;

namespace GameEventScript.Conformance;

internal static class ConformanceRuntimeValueCodec
{
    internal static GameEventScriptMessage DecodeMessage(ConformanceMessage message)
    {
        var arguments = new GameEventScriptMessageArgument[message.Arguments.Count];
        for (var index = 0; index < arguments.Length; index++)
            arguments[index] = new GameEventScriptMessageArgument(message.Arguments[index].Name, DecodeValue(message.Arguments[index].Value));
        return GameEventScriptMessage.Create(message.Name, arguments, message.Tags);
    }

    internal static GesValue DecodeValue(ConformanceValue value)
    {
        var unit = ParseUnit(value.Unit);
        switch (value.Type)
        {
            case ":Nothing": return GesValue.GesNothing();
            case ":Text": return GesValue.GesText(value.Value ?? string.Empty);
            case ":Tag": return GesValue.GesTag(value.Value ?? string.Empty);
            case ":Boolean": return GesValue.GesBoolean(string.Equals(value.Value, "true", StringComparison.Ordinal));
            case ":Number.int64": return GesValue.GesInteger(long.Parse(value.Value!, NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture));
            case ":Number.binary64": return GesValue.GesFloat(ParseBinary64(value.Value!));
            case ":Quantity.int64": return GesValue.GesInteger(long.Parse(value.Value!, NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture), unit);
            case ":Quantity.binary64": return GesValue.GesFloat(ParseBinary64(value.Value!), unit);
            case ":Percentage": return GesValue.GesPercentage(ParseBinary64(value.Value!));
            case ":Vector": return GesValue.GesVector(ParseBinary64(value.X!), ParseBinary64(value.Y!), ParseBinary64(value.Z!), unit);
            case ":Point": return GesValue.GesPoint(ParseBinary64(value.X!), ParseBinary64(value.Y!), ParseBinary64(value.Z!), unit);
            case ":List":
            {
                var items = new GesValue[value.Items.Count];
                for (var index = 0; index < items.Length; index++) items[index] = DecodeValue(value.Items[index]);
                return GesValue.GesList(items);
            }
            case ":Map": return DecodeMap(value.Entries);
            case ":Dice":
            {
                var rolls = new int[value.Rolls.Count];
                for (var index = 0; index < rolls.Length; index++) rolls[index] = value.Rolls[index];
                return GesValue.GesDice(rolls);
            }
            case ":Range.int64":
                return GesValue.GesRange(
                    long.Parse(value.From!, NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture),
                    long.Parse(value.To!, NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture),
                    long.Parse(value.Step!, NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture));
            case ":Range.binary64": return GesValue.GesRange(ParseBinary64(value.From!), ParseBinary64(value.To!), ParseBinary64(value.Step!));
            case ":Message": return GesValue.GesMessage(DecodeMessage(value.Message!));
            default:
            {
                var keys = new string[value.Entries.Count];
                var values = new GesValue[value.Entries.Count];
                for (var index = 0; index < keys.Length; index++)
                {
                    keys[index] = value.Entries[index].Key;
                    values[index] = DecodeValue(value.Entries[index].Value);
                }
                return GesValue.GesRecord(value.Type.Substring(1), keys, values);
            }
        }
    }

    internal static GesValue DecodeValueAndMutateSource(ConformanceValue value)
    {
        switch (value.Type)
        {
            case ":List":
            {
                var items = new GesValue[value.Items.Count];
                for (var index = 0; index < items.Length; index++) items[index] = DecodeValue(value.Items[index]);
                var result = GesValue.GesList(items);
                if (items.Length > 0) items[0] = GesValue.GesNothing();
                return result;
            }
            case ":Map":
            {
                var keys = new string[value.Entries.Count];
                var values = new GesValue[value.Entries.Count];
                for (var index = 0; index < keys.Length; index++)
                {
                    keys[index] = value.Entries[index].Key;
                    values[index] = DecodeValue(value.Entries[index].Value);
                }
                var result = GesValue.GesMap(keys, values);
                if (keys.Length > 0) { keys[0] = "mutated"; values[0] = GesValue.GesNothing(); }
                return result;
            }
            case ":Dice":
            {
                var rolls = new int[value.Rolls.Count];
                for (var index = 0; index < rolls.Length; index++) rolls[index] = value.Rolls[index];
                var result = GesValue.GesDice(rolls);
                if (rolls.Length > 0) rolls[0] = int.MinValue;
                return result;
            }
            case var type when type.Length > 1 &&
                                   type is not ":Nothing" and not ":Text" and not ":Tag" and not ":Boolean" and
                                   not ":Number.int64" and not ":Number.binary64" and not ":Quantity.int64" and not ":Quantity.binary64" and
                                   not ":Percentage" and not ":Vector" and not ":Point" and not ":Range.int64" and not ":Range.binary64" and not ":Message":
            {
                var keys = new string[value.Entries.Count];
                var values = new GesValue[value.Entries.Count];
                for (var index = 0; index < keys.Length; index++)
                {
                    keys[index] = value.Entries[index].Key;
                    values[index] = DecodeValue(value.Entries[index].Value);
                }
                var result = GesValue.GesRecord(value.Type.Substring(1), keys, values);
                if (keys.Length > 0) { keys[0] = "mutated"; values[0] = GesValue.GesNothing(); }
                return result;
            }
            default: return DecodeValue(value);
        }
    }

    private static GesValue DecodeMap(IReadOnlyList<ConformanceValueEntry> entries)
    {
        var keys = new string[entries.Count];
        var values = new GesValue[entries.Count];
        for (var index = 0; index < keys.Length; index++)
        {
            keys[index] = entries[index].Key;
            values[index] = DecodeValue(entries[index].Value);
        }
        return GesValue.GesMap(keys, values);
    }

    internal static bool MessagesEqual(ConformanceMessage expected, GameEventScriptMessage actual, ConformanceComparisonOptions comparison)
    {
        if (!string.Equals(expected.Name, actual.Name, StringComparison.Ordinal) || expected.Tags.Count != actual.Tags.Count || expected.Arguments.Count != actual.Arguments.Count) return false;
        for (var index = 0; index < expected.Tags.Count; index++)
            if (!string.Equals(expected.Tags[index], actual.Tags[index], StringComparison.Ordinal)) return false;
        for (var index = 0; index < expected.Arguments.Count; index++)
        {
            if (!string.Equals(expected.Arguments[index].Name, actual.Arguments.NameAt(index), StringComparison.Ordinal)) return false;
            var actualValue = actual.Arguments.ValueAt(index);
            if (!ValuesEqual(expected.Arguments[index].Value, in actualValue, comparison)) return false;
        }
        return true;
    }

    internal static bool ValuesEqual(ConformanceValue expected, in GesValue actual, ConformanceComparisonOptions comparison)
    {
        // Expected transport values must not pass through factories that normalize storage kinds.
        // NaN is the explicitly supported conformance spelling of invalid mathematics (nothing).
        var kind = actual.ValueKind;
        switch (expected.Type)
        {
            case ":Nothing": return actual.IsNothing;
            case ":Boolean": return kind == GameEventScriptBytecodeTypeKind.Boolean && (expected.Value == "true") == actual.AsBoolean();
            case ":Number.int64":
            case ":Quantity.int64":
                return kind == Integer && ParseUnit(expected.Unit) == actual.ValueUnit && ParseInt64(expected.Value!) == actual.AsInteger();
            case ":Number.binary64":
            case ":Quantity.binary64":
            case ":Percentage":
            {
                var number = ParseBinary64(expected.Value!);
                if (double.IsNaN(number)) return actual.IsNothing;
                return kind == (expected.Type == ":Percentage" ? Percentage : Float) && ParseUnit(expected.Unit) == actual.ValueUnit && Binary64Equal(number, actual.AsNumber(), comparison);
            }
            case ":Text": return kind == Text && string.Equals(expected.Value, actual.AsText(), StringComparison.Ordinal);
            case ":Tag": return kind == Tag && string.Equals(expected.Value, actual.AsText(), StringComparison.Ordinal);
            case ":Vector":
            case ":Point":
            {
                var x = ParseBinary64(expected.X!);
                var y = ParseBinary64(expected.Y!);
                var z = ParseBinary64(expected.Z!);
                if (double.IsNaN(x) || double.IsNaN(y) || double.IsNaN(z)) return actual.IsNothing;
                return kind == (expected.Type == ":Vector" ? Vector : Point) && ParseUnit(expected.Unit) == actual.ValueUnit &&
                       Binary64Equal(x, actual.X, comparison) && Binary64Equal(y, actual.Y, comparison) && Binary64Equal(z, actual.Z, comparison);
            }
            case ":List":
            {
                if (kind != List) return false;
                var right = actual.AsList();
                if (expected.Items.Count != right.Length) return false;
                for (var index = 0; index < right.Length; index++)
                {
                    var rightValue = right[index];
                    if (!ValuesEqual(expected.Items[index], in rightValue, comparison)) return false;
                }
                return true;
            }
            case ":Map": return kind == Map && MapEntriesEqual(expected.Entries, actual.AsMap(), comparison);
            case ":Dice":
            {
                if (kind != Dice) return false;
                var right = actual.AsDice();
                if (expected.Rolls.Count != right.Length) return false;
                for (var index = 0; index < right.Length; index++) if (expected.Rolls[index] != right[index]) return false;
                return true;
            }
            case ":Range.int64":
                return kind == GameEventScriptBytecodeTypeKind.Range && actual.IntegerRange is { } integerRange &&
                       ParseInt64(expected.From!) == integerRange.From && ParseInt64(expected.To!) == integerRange.To && ParseInt64(expected.Step!) == integerRange.Step;
            case ":Range.binary64":
            {
                var from = ParseBinary64(expected.From!);
                var to = ParseBinary64(expected.To!);
                var step = ParseBinary64(expected.Step!);
                if (double.IsNaN(from) || double.IsNaN(to) || double.IsNaN(step)) return actual.IsNothing;
                return kind == GameEventScriptBytecodeTypeKind.Range && actual.FloatRange is { } floatRange && Binary64Equal(from, floatRange.From, comparison) &&
                       Binary64Equal(to, floatRange.To, comparison) && Binary64Equal(step, floatRange.Step, comparison);
            }
            case ":Message": return kind == Message && expected.Message is not null && actual.Message is not null && MessagesEqual(expected.Message, actual.Message, comparison);
            default:
                return kind == Custom && string.Equals(expected.Type.Substring(1), actual.CustomTypeName, StringComparison.Ordinal) && MapEntriesEqual(expected.Entries, actual.AsMap(), comparison);
        }
    }

    private static bool MapEntriesEqual(IReadOnlyList<ConformanceValueEntry> expected, GesValueMap? actual, ConformanceComparisonOptions comparison)
    {
        if (actual is null) return false;
        var entries = new Dictionary<string, ConformanceValue>(StringComparer.Ordinal);
        for (var index = 0; index < expected.Count; index++) entries[expected[index].Key] = expected[index].Value;
        if (entries.Count != actual.Length) return false;
        foreach (var entry in entries)
        {
            var value = actual.Get(entry.Key);
            if (value is not { } right || !ValuesEqual(entry.Value, in right, comparison)) return false;
        }
        return true;
    }

    private static long ParseInt64(string value) => long.Parse(value, NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture);

    internal static string Describe(GameEventScriptMessage message) => message.ToString();

    private static GameEventScriptBytecodeInstructionUnit ParseUnit(string? unit)
        => unit?.TrimStart(':') switch
        {
            null => GameEventScriptBytecodeInstructionUnit.UnitNone,
            "m" => GameEventScriptBytecodeInstructionUnit.UnitMeter,
            "s" => GameEventScriptBytecodeInstructionUnit.UnitSecond,
            var name => GameEventScriptBytecodeInstructionUnits.ParseTypeName(name) ?? GameEventScriptBytecodeInstructionUnit.UnitInvalid
        };

    internal static double ParseBinary64(string value) => value switch
    {
        "NaN" => double.NaN,
        "Infinity" => double.PositiveInfinity,
        "-Infinity" => double.NegativeInfinity,
        _ => double.Parse(value, NumberStyles.Float, CultureInfo.InvariantCulture)
    };

    internal static string FormatBinary64(double value) => value.ToString("R", CultureInfo.InvariantCulture);

    private static bool Binary64Equal(double left, double right, ConformanceComparisonOptions comparison)
    {
        if (double.IsNaN(left) || double.IsNaN(right)) return double.IsNaN(left) && double.IsNaN(right);
        if (comparison.Binary64Mode == ConformanceBinary64ComparisonMode.Exact)
            return BitConverter.DoubleToInt64Bits(left) == BitConverter.DoubleToInt64Bits(right);
        if (left == right) return true;
        if (double.IsInfinity(left) || double.IsInfinity(right)) return false;
        var leftOrdered = OrderedBits(left);
        var rightOrdered = OrderedBits(right);
        var distance = leftOrdered >= rightOrdered ? leftOrdered - rightOrdered : rightOrdered - leftOrdered;
        return distance <= comparison.MaxUlps;
    }

    private static ulong OrderedBits(double value)
    {
        var bits = unchecked((ulong)BitConverter.DoubleToInt64Bits(value));
        return (bits & 0x8000000000000000UL) != 0 ? ~bits : bits | 0x8000000000000000UL;
    }
}
